<a href="https://colab.research.google.com/github/SusanaAlvarezC/Fundamentos_Clases/blob/main/Prueba2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Estudiantes**:

*  ## Alejandro Martínez Hernández | alemaher12@gmail.com
*  ## Daniel Marín Vélez           | daniel.marinv@udea.edu.co
*  ## Susana Álvarez               | susanamalc@gmail.com  

   
### ***Nota para el profesor:***

**↤ ↤** Puede usar el navegador a la izquierda **↤ ↤**

# **Prueba \#2**




## **Ejercicio 1** (40 pts)

Construya un DataFrame a partir del siguiente diccionario

exam_data = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'], 'score': [12.5, 9, 16.5, 10, 9, 20, 14.5, 12, 8, 19], 'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1], 'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}

con las siguientes etiquetas para los índices

labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

In [ ]:
# Construir el DataFrame
import numpy as np
import pandas as pd

# Almacenamos la información en un diccionario y el índice en una lista
# para posteriormente crear el DataFrame usando la librería Pandas
exam_data = {'name': ['Anastasia', 'Dima', 'Katherine', 'James',
                      'Emily','Michael', 'Matthew', 'Laura', 'Kevin',
                      'Jonas'],
             'score': [12.5, 9, 16.5, 10, 9, 20, 14.5, 12, 8, 19],
             'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
             'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes',
                         'no', 'no', 'yes']}

labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

df_exam = pd.DataFrame(exam_data, index = labels)

# Somos conscientes que en otro editor de texto debemos imprimir
# nuestro Output, pero para esta prueba solo escribiremos la variable
# para que se nos muestre en pantalla.
df_exam

name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes
d      James   10.0         3      no
e      Emily    9.0         2      no
f    Michael   20.0         3     yes
g    Matthew   14.5         1     yes
h      Laura   12.0         1      no
i      Kevin    8.0         2      no
j      Jonas   19.0         1     yes

### **1.1. Seleccionar columnas name y score**

In [ ]:
# Con el método .loc() seleccionamos todas las filas
# de las columnas 'name' y 'score' y almacenamos en df_1
df_1 = df_exam.loc[:, ['name', 'score']]

df_1

name  score
a  Anastasia   12.5
b       Dima    9.0
c  Katherine   16.5
d      James   10.0
e      Emily    9.0
f    Michael   20.0
g    Matthew   14.5
h      Laura   12.0
i      Kevin    8.0
j      Jonas   19.0

### **1.2. Filtre las filas donde el número de intentos (attemps) sea mayor que 2**

In [ ]:
# Nuevamente usamos .loc(), pero para solo seleccionar las filas que
# cumplan la condición.
df_2 = df_exam.loc[df_exam['attempts']>2]

df_2

name  score  attempts qualify
b     Dima    9.0         3      no
d    James   10.0         3      no
f  Michael   20.0         3     yes

### **1.3. Filtre las filas donde el score se encuentre entre 15 y 20 (incluidos)**

In [ ]:
# A través del método .loc() agregamos las condiciones
# deseadas y almacenamos resultados en df_3 y ordenamos
df_3 = df_exam.loc[(df_exam.score >=15) & (df_exam.score <=20)]\
.sort_values(by='score')

df_3

name  score  attempts qualify
c  Katherine   16.5         2     yes
j      Jonas   19.0         1     yes
f    Michael   20.0         3     yes

### **1.4. Cambie el score en la fila 'd' a 11.5**

In [ ]:
# Usamos el método .replace() especificando los valores a reemplazar
df_4 = df_exam.replace([10], 11.5)

# Solo mostramos las primeras 5 filas para facilitar lectura
df_4.head(5)

name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes
d      James   11.5         3      no
e      Emily    9.0         2      no

### **1.5. Agregue una nueva fila con etiqueta 'k' al dataframe con un valor que usted desee para cada columna**

In [ ]:
# Creamos una copia del DataFrame para no modificarlo
# A la copia asignamos una nueva fila con valores inventados
df_5 = df_exam.copy()
df_5.loc['k'] = ['Coraline', 19.0, 3, 'yes']

df_5

name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes
d      James   10.0         3      no
e      Emily    9.0         2      no
f    Michael   20.0         3     yes
g    Matthew   14.5         1     yes
h      Laura   12.0         1      no
i      Kevin    8.0         2      no
j      Jonas   19.0         1     yes
k   Coraline   19.0         3     yes

### **1.6. Borre la fila 'k' y entregue nuevamente el DataFrame original.**

In [ ]:
# Usando la copia creada en el punto 1.5
# y el método .drop(), eliminamos la fila deseada.
df_6 = df_5.drop('k')

df_6

name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes
d      James   10.0         3      no
e      Emily    9.0         2      no
f    Michael   20.0         3     yes
g    Matthew   14.5         1     yes
h      Laura   12.0         1      no
i      Kevin    8.0         2      no
j      Jonas   19.0         1     yes

### **1.7. Ordene DataFrame por nombre ascendentemente**

In [ ]:
# Se realiza a través del método .sort_values()
df_7 = df_exam.sort_values(by=['name'], ascending = True)

df_7

name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
e      Emily    9.0         2      no
d      James   10.0         3      no
j      Jonas   19.0         1     yes
c  Katherine   16.5         2     yes
i      Kevin    8.0         2      no
h      Laura   12.0         1      no
g    Matthew   14.5         1     yes
f    Michael   20.0         3     yes

### **1.8. Reemplace los valores de la columna qualify que contiene los valores yes y no por los valores booleanos True y False respectivamente.**

In [ ]:
# Se reliza a través del método .replace()
df_8 = df_exam.replace({'qualify' : {'yes':True, 'no':'False'}})

df_8

name  score  attempts qualify
a  Anastasia   12.5         1    True
b       Dima    9.0         3   False
c  Katherine   16.5         2    True
d      James   10.0         3   False
e      Emily    9.0         2   False
f    Michael   20.0         3    True
g    Matthew   14.5         1    True
h      Laura   12.0         1   False
i      Kevin    8.0         2   False
j      Jonas   19.0         1    True

## **Ejercicio 2** (30 %)

Importe el dataset athletes del repositorio del curso y realice las siguientes tareas:

1. Importarlo como nombre athletes
2. Filtre el dataframe tal que solo se quede con los atletas que hayan ganado al menos una medalla.
3. Calcule la masa muscular por fila con una función como

4. Agrupe los datos por calculando la media por deporte para cada una del resto de características. ¿Cuál es el deporte en el que los jugadores tienen mayor peso promedio?

5. ¿Cuál es el pais con más participación femenina?


### **2.1. Importarlo como nombre athletes**


In [ ]:
import pandas as pd

athletes = pd.read_csv("https://raw.githubusercontent.com/tomasate/Datos_Clases/main/Datos_1/athletes.csv")
athletes.head()

id            name nationality     sex       dob  height  weight  \
0  736041664  A Jesus Garcia         ESP    male  10/17/69    1.72    64.0   
1  532037425      A Lam Shin         KOR  female   9/23/86    1.68    56.0   
2  435962603     Aaron Brown         CAN    male   5/27/92    1.98    79.0   
3  521041435      Aaron Cook         MDA    male    1/2/91    1.83    80.0   
4   33922579      Aaron Gate         NZL    male  11/26/90    1.81    71.0   

       sport  gold  silver  bronze  
0  athletics     0       0       0  
1    fencing     0       0       0  
2  athletics     0       0       1  
3  taekwondo     0       0       0  
4    cycling     0       0       0

### **2.2. Filtre el dataframe tal que solo se quede con los atletas que hayan ganado al menos una medalla.**

In [ ]:
# Filtrar el DataFrame para quedarse solo con los atletas que tienen
# un total de medallas mayor o igual a 1
filtered_df = athletes[(athletes['gold'] >= 1) |
 (athletes['silver'] >= 1) | (athletes['bronze'] >= 1) ]

filtered_df

id                         name nationality     sex       dob  \
2      435962603                  Aaron Brown         CAN    male   5/27/92   
6      266237702                Aaron Russell         USA    male    6/4/93   
14     162792594               Abbey Weitzeil         USA  female   12/3/96   
48     962468808  Abdoulrazak Issoufou Alfaga         NIG    male  12/26/94   
55     969824503           Abdullah Alrashidi         IOA    male   8/21/63   
...          ...                          ...         ...     ...       ...   
11487  334126873                  Zhiyong Shi         CHN    male  10/10/93   
11491  105609876                        Zi He         CHN  female  12/10/90   
11503  593183472                  Zivko Gocic         SRB    male   8/22/82   
11509  836633838                      Zoe Lee         GBR  female  12/15/85   
11511  343671816                Zoe de Toledo         GBR  female   7/17/87   

       height  weight          sport  gold  silver  bronze  
2        1.98    79.0      athletics     0       0       1  
6        2.05    98.0     volleyball     0       0       1  
14       1.78    68.0       aquatics     1       1       0  
48       2.07    98.0      taekwondo     0       1       0  
55       1.83    84.0       shooting     0       0       1  
...       ...     ...            ...   ...     ...     ...  
11487    1.68    69.0  weightlifting     1       0       0  
11491    1.60    52.0       aquatics     0       1       0  
11503    1.93    93.0       aquatics     1       0       0  
11509    1.76    70.0         rowing     0       1       0  
11511    1.72    58.0         rowing     0       1       0  

[1857 rows x 11 columns]

### **2.3. Calcule la masa muscular por fila con una función como**

$$I_{m.m.} = \frac{Weight}{{Height}^2}$$

In [ ]:
def imm(row, variable, variable2):

  """
    Calcula el índice de masa muscular (IMM) utilizando los valores de
    las columnas 'variable' y 'variable2' en la fila.

    Args:
        row (pandas.Series): Fila del DataFrame con los valores a
            utilizar para el cálculo.
        variable (str): Nombre de la columna que contiene el valor
            del numerador (peso).
        variable2 (str): Nombre de la columna que contiene el valor
            del denominador al cuadrado (altura).

    Returns:
        float or None: El valor calculado del índice de masa muscular
            o None si el denominador es cero.

    """
  if(row.loc[variable2]>0):
    return row.loc[variable]/row.loc[variable2]**2


"""
Aplicar la función imm a cada fila del DataFrame 'athletes' utilizando
apply con axis=1. Los argumentos 'weight' y 'height' se pasan a la
función como argumentos. El resultado se agrega como una nueva columna
llamada 'Imm' en el DataFrame.
"""
athletes['Imm'] = athletes.apply(imm, axis = 1,
                                 args =['weight', 'height'])

athletes

id                name nationality     sex       dob  height  \
0      736041664      A Jesus Garcia         ESP    male  10/17/69    1.72   
1      532037425          A Lam Shin         KOR  female   9/23/86    1.68   
2      435962603         Aaron Brown         CAN    male   5/27/92    1.98   
3      521041435          Aaron Cook         MDA    male    1/2/91    1.83   
4       33922579          Aaron Gate         NZL    male  11/26/90    1.81   
...          ...                 ...         ...     ...       ...     ...   
11533  265605954  Zurian Hechavarria         CUB  female   8/10/95    1.64   
11534  214461847      Zuzana Hejnova         CZE  female  12/19/86    1.73   
11535   88361042             di Xiao         CHN    male   5/14/91    1.85   
11536  900065925   le Quoc Toan Tran         VIE    male    4/5/89    1.60   
11537  711404576      le Roux Hamman         RSA    male    1/6/92    1.85   

       weight          sport  gold  silver  bronze        Imm  
0        64.0      athletics     0       0       0  21.633315  
1        56.0        fencing     0       0       0  19.841270  
2        79.0      athletics     0       0       1  20.151005  
3        80.0      taekwondo     0       0       0  23.888441  
4        71.0        cycling     0       0       0  21.672110  
...       ...            ...   ...     ...     ...        ...  
11533    58.0      athletics     0       0       0  21.564545  
11534    63.0      athletics     0       0       0  21.049818  
11535   100.0      wrestling     0       0       0  29.218408  
11536    56.0  weightlifting     0       0       0  21.875000  
11537    70.0      athletics     0       0       0  20.452885  

[11538 rows x 12 columns]

### **2.4. Agrupando deporte por media**

#### **2.4.1 Media por deporte**
Agrupe los datos por calculando la media por deporte para cada una del resto de características.

In [ ]:
athletes.groupby(['sport'])[['height','weight','gold','silver','bronze']].mean()

height     weight      gold    silver    bronze
sport                                                               
aquatics           1.788338  72.301644  0.083045  0.077509  0.073356
archery            1.735952  72.190476  0.062500  0.062500  0.062500
athletics          1.751922  67.717727  0.027931  0.027084  0.026238
badminton          1.745529  68.774390  0.046512  0.046512  0.046512
basketball         1.918715  87.750000  0.083333  0.083333  0.083333
boxing             1.747745        NaN  0.045455  0.045455  0.090909
canoe              1.781006  77.015291  0.081571  0.081571  0.084592
cycling            1.746667  67.820717  0.051429  0.053333  0.055238
equestrian         1.748704  67.493023  0.067568  0.067568  0.067568
fencing            1.772449  70.661224  0.085366  0.089431  0.089431
football           1.745341  68.433962  0.058920  0.058920  0.058920
golf               1.743761  71.443478  0.016667  0.016667  0.016667
gymnastics         1.637840  54.278997  0.092593  0.092593  0.092593
handball           1.837080  83.710602  0.079890  0.082645  0.079890
hockey             1.733981  68.900463  0.078704  0.074074  0.076389
judo               1.735366  76.876316  0.035714  0.035714  0.071429
modern pentathlon  1.762361  65.958333  0.027778  0.027778  0.027778
rowing             1.849477  79.938318  0.087751  0.087751  0.087751
rugby sevens       1.755623  78.723906  0.083333  0.080000  0.083333
sailing            1.761478  71.169355  0.039474  0.039474  0.039474
shooting           1.728668  73.905263  0.038462  0.038462  0.038462
table tennis       1.721845  65.177515  0.046512  0.046512  0.046512
taekwondo          1.797480  68.088000  0.062500  0.062500  0.125000
tennis             1.802251  73.162304  0.040816  0.040816  0.040816
triathlon          1.732110  60.633028  0.018182  0.018182  0.018182
volleyball         1.896527  80.102094  0.072917  0.070312  0.072917
weightlifting      1.666628  79.980620  0.058140  0.058140  0.058140
wrestling          1.723000  77.742120  0.050992  0.050992  0.101983

#### **2.4.1 Mayor peso promedio**

¿Cuál es el deporte en el que los jugadores tienen mayor peso promedio?

##### **Respuesta:**
***El deporte en que los jugadores tienen mayor peso promedio es basketball con un peso promedio de 87.75***

In [ ]:
group_weight_sport = athletes.groupby(['sport'])[['weight']].mean()
group_weight_sport.sort_values(by='weight', ascending=False).iloc[0]

weight    87.75
Name: basketball, dtype: float64

### **2.5. País con mayor participación femenina**

#### **Respuesta:**
***El país con mayor participación femenina es USA con 303 candidatas***

In [ ]:
# Filtramos para obtener atletas de sex = 'Female', agrupamos por
# nacionalidad y sexo y finalmente contamos:
group = athletes[(athletes['sex'] == 'female')]\
.groupby(['nationality'])[['sex']].count()

# Ordenamos la cantidad descendentemente y mostramos solo el primer
# registro:
group.sort_values(by='sex',ascending=False).iloc[:1]


sex
nationality     
USA          303

## **Ejercicio 3** 30%

Importe el dataset de nba del repositorio de clase y responda (hint: Elimine las instancias con registros faltantes):

1. Cuáles son los equipos que pagan los salarios más altos (presente el resultado como un dataFrame con el nombre del equipo y la cantidad de dinero de la nómina)
2. Separe la variable Age en tres grupos y cree una nueva columna llamada Age-Group con esta información.<
3. Construya una tabla dinámica que muestre el salario promedio de los jugadores por equipo y posición para para los cinco equipos que pagan mejores salarios.



In [ ]:
import numpy as np
import pandas as pd

# Con la función pd.read se extraen los datos que están generados en el
# archivo nba
nba = pd.read_csv(
    'https://raw.githubusercontent.com/tomasate/Datos_Clases/main/Datos_1/nba.csv'
    ,index_col=0)

# Con index_col se salta la primera columna que no tiene informacion
# relevante
nba.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

In [ ]:
# El método.dropna() eliminará todas las filas que contengan
# al menos un valor faltante (NaN) en cualquier columna.
nba_filtered = nba.dropna()

nba_filtered.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
6  Jordan Mickey  Boston Celtics    55.0       PF  21.0    6-8   235.0   
7   Kelly Olynyk  Boston Celtics    41.0        C  25.0    7-0   238.0   

         College     Salary  
0          Texas  7730337.0  
1      Marquette  6796117.0  
3  Georgia State  1148640.0  
6            LSU  1170960.0  
7        Gonzaga  2165160.0

### **3.1. Equipos con salarios mas altos**
Cuáles son los equipos que pagan los salarios más altos (presente el resultado como un dataFrame con el nombre del equipo y la cantidad de dinero de la nómina)

In [ ]:
df_highest_salary = nba_filtered.groupby(['Team'])[['Salary']].sum()

df_highest_salary_sorted = df_highest_salary.sort_values(
    by='Salary', ascending=False)

# Cambiamos el nombre de la columna 'id' a 'cantidad'
df_highest_salary_sorted = df_highest_salary_sorted.rename(
    columns={'Team': 'Nombre del Equipo', 'Salary': 'Cantidad Dinero'})

# Solicitar al usuario el número de filas que desea visualizar
top_n = int(input('Escriba el top de salarios que desea visualizar:'))

# Mostrar las primeras n filas según la entrada del usuario
df_highest_salary_sorted.head(top_n)

Escriba el top de salarios que desea visualizar:4


Cantidad Dinero
Team                                  
Los Angeles Clippers        93906914.0
Oklahoma City Thunder       81515298.0
Golden State Warriors       80535517.0
Cleveland Cavaliers         74068189.0

### **3.2. Separe la variable Age en tres grupos**



#### 3.2.1. Separe la variable Age en tres grupos y cree una nueva columna llamada Age-Group con esta información

In [ ]:
nba_filtered['Age-Group'] = pd.cut(
    nba_filtered['Age'], bins=3, labels=[
        'Primer Grupo', 'Segundo Grupo', 'Tercer grupo'])

nba_filtered

Name            Team  Number Position   Age Height  Weight  \
0    Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1      Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
3      R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
6    Jordan Mickey  Boston Celtics    55.0       PF  21.0    6-8   235.0   
7     Kelly Olynyk  Boston Celtics    41.0        C  25.0    7-0   238.0   
..             ...             ...     ...      ...   ...    ...     ...   
449    Rodney Hood       Utah Jazz     5.0       SG  23.0    6-8   206.0   
451  Chris Johnson       Utah Jazz    23.0       SF  26.0    6-6   206.0   
452     Trey Lyles       Utah Jazz    41.0       PF  20.0   6-10   234.0   
453   Shelvin Mack       Utah Jazz     8.0       PG  26.0    6-3   203.0   
456    Jeff Withey       Utah Jazz    24.0        C  26.0    7-0   231.0   

           College     Salary     Age-Group  
0            Texas  7730337.0  Primer Grupo  
1        Marquette  6796117.0  Primer Grupo  
3    Georgia State  1148640.0  Primer Grupo  
6              LSU  1170960.0  Primer Grupo  
7          Gonzaga  2165160.0  Primer Grupo  
..             ...        ...           ...  
449           Duke  1348440.0  Primer Grupo  
451         Dayton   981348.0  Primer Grupo  
452       Kentucky  2239800.0  Primer Grupo  
453         Butler  2433333.0  Primer Grupo  
456         Kansas   947276.0  Primer Grupo  

[364 rows x 10 columns]

#### 3.2.2. Presentamos la edad mínima y máxima que se asignó a cada grupo

In [ ]:
no nba_filtered.groupby('Age-Group')['Age'].agg(['min', 'max'])

min   max
Age-Group                
Primer Grupo   19.0  26.0
Segundo Grupo  27.0  33.0
Tercer grupo   34.0  40.0

### **3.3. Tabla dinamica salario por equipo y posición**
Construya una tabla dinámica que muestre el salario promedio de los jugadores por equipo y posición para para los cinco equipos que pagan mejores salarios.

In [ ]:
# Primero calculamos el promedio del salario por equipo y posicion,
# con sort_Values ordenamos por salario descendente y con head tomamos
# los primeros 5 valores
avg_salary = nba_filtered.groupby(['Team', 'Position'])[['Salary']]\
.mean().sort_values(by = 'Salary',ascending = False).head(5)

avg_salary

Salary
Team                  Position            
Golden State Warriors PF        14260870.0
Cleveland Cavaliers   PF        13941014.5
Chicago Bulls         C         13400000.0
New York Knicks       C         12650000.0
Los Angeles Clippers  PG        12289745.5

Creamos tabla dinamica con el resultado anterior

In [ ]:
pivot_table = pd.pivot_table(avg_salary, values='Salary',
                             index='Team', columns='Position',
                             fill_value=0)

pivot_table

Position                      C          PF          PG
Team                                                   
Chicago Bulls          13400000         0.0         0.0
Cleveland Cavaliers           0  13941014.5         0.0
Golden State Warriors         0  14260870.0         0.0
Los Angeles Clippers          0         0.0  12289745.5
New York Knicks        12650000         0.0         0.0